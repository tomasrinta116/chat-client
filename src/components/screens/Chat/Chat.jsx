import React from "react";
import CircularProgress from "@material-ui/core/CircularProgress";
import { withStyles } from "@material-ui/core/styles";
import ChatHeader from "./ChatHeader/ChatHeader";
import ChatMenu from "./ChatMenu/ChatMenu";
import NavBar from "../../common/NavBar/NavBar";
import styles from "./Chat.styles";
import MessageInput from "./MessageInput/MessageInput";
import MessagesDisplay from "./MessagesDisplay/MessagesDisplay";
import { addChannelHandler, getParticipantList } from "../../../utils/channelHandler";
import { getChannel, exitChannel, getMessages } from "../../../utils/sendbirdHelpers";
import PropTypes from 'prop-types';

class Chat extends React.Component {
  state = {
    loading: true,
    channel: "",
    messages: [],
    channelName: "",
    participants: []
  };

  async componentDidMount() {
    const { sb } = this.props;
    const channelURL = this.props.match.params.channelURL;
    const channel = await getChannel(sb, channelURL);
    let initialParticipants = getParticipantList(channel);
    let prevMessages = await getMessages(channel);
    prevMessages = prevMessages.map(message => {
      return this.transformMessage(message);
    })
    window.addEventListener("beforeunload", this.onUnload);
    this.setState({
      channel: channel,
      channelName: channel.name,
      messages: prevMessages,
      participants: initialParticipants,
      loading: false
    });
    addChannelHandler(sb, channel, this.updateParticipants, this.addNewMessage);
  }

  componentWillUnmount() {
    window.removeEventListener("beforeunload", this.onUnload);
  }

  onUnload = event => {
    event.preventDefault();
    exitChannel(this.props.sb, this.state.channel); // this might need to be logout
    // Chrome requires returnValue to be set
    event.returnValue = "";
  };

  updateParticipants = participantsList => {
    this.setState({ participants: participantsList });
  };

  addNewMessage = newMessage => {
    let transformedMessage = this.transformMessage(newMessage);
    this.setState({
      messages: [...this.state.messages, transformedMessage]
    });
  };

  transformMessage = (messageObj) => {
    const {sb} = this.props;
    const messageContent = messageObj.message;
    const userName = sb.currentUser.userId;
    const senderName = messageObj._sender.userId;
    // This check is needed when loading previous messages, 
    // as they won't be tagged with "You" for the senderName.
    if (senderName === userName) { 
      return { sender: "You", message: messageContent}
    }
    return { sender: senderName, message: messageContent }
  }

  render() {
    const {
      classes: {
        loadingSpinner,
        chatHeader,
        chatContainer,
        messagesDisplay,
        createMessage
      },
      history,
      sb,
      match
    } = this.props;
    const {
      channelName,
      messages,
      loading,
      channel,
      participants
    } = this.state;

    if (loading === true) {
      return (
        <div className={loadingSpinner}>
          <CircularProgress />
        </div>
      );
    }

    return (
      <React.Fragment>
        <NavBar history={history} sb={sb} match={match} />
        <div className={chatContainer}>
          <div className={chatHeader}>
            <ChatHeader title={channelName}>
              <ChatMenu
                history={history}
                channel={channel}
                participants={participants}
                sb={sb}
              />
            </ChatHeader>
          </div>
          <div className={messagesDisplay}>
            <MessagesDisplay messages={messages} />
          </div>
          <div className={createMessage}>
            <MessageInput channel={channel} addMessage={this.addNewMessage} />
          </div>
        </div>
      </React.Fragment>
    );
  }
}

Chat.propTypes = {
  sb: PropTypes.object.isRequired,
}

export default withStyles(styles)(Chat);
