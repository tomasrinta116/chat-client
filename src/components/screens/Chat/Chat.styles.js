const styles = {
  chatContainer: {
    width: "100%",
    height: "90vh", // maybe position at bottom?
    display: "flex",
    flexFlow: "column",
    justifyContent: "flex-end"
  },
  createMessage: {
    width: "100%"
  },
  chatHeader: {
    maxHeight: "7vh",
    minHeight: "7vh",
    marginBottom: "auto",
    display: "flex",
    flexFlow: "row",
    alignItems: "center",
    paddingLeft: "10px",
    paddingRight: "10px"
  },
  loadingSpinner: {
    height: "90vh", // maybe position at bottom?
    display: "flex",
    alignItems: "center",
    justifyContent: "center"
  },
  messagesDisplay: {
    overflowY: "auto"
  }
};

export default styles;
