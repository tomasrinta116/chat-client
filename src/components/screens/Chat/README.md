# Chat Overview

This folder contains all the contents for the Chat screen. The primary parent component - <Chat> - contains the majority of the data and transformation logic that is fed into the other components. Many of the functions and procedures used are built off of functionality provided by the SendBird SDK. If this functionality is used in multiple places, it has been abstracted into the "utils" folder. Of particular importance is the ChannelHandler/ listener functionality. When the chat component mounts with a channel assigned, a channel handler is added to it that listens to activity from that channel (messages received, user joined/ left, etc.). This functionality is largely served by the channelHandler.js file within the utils folder. 