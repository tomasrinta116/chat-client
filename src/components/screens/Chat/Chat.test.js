import React from "react";
import { shallow } from "enzyme";
import Chat from './Chat';

describe("<Chat />", () => {
    let wrapper;
    const props = {
        channel: {name: 'test-channel'}, 
        sb: {},
        history: {},
    }

    beforeEach(() => {
        wrapper = shallow(<Chat {...props} />).dive();
    })

    test("it renders a loading spinner while loading", () => {
        expect(wrapper).toMatchSnapshot();
    })

    test("it renders after loading", () => {
        wrapper.setState({ 
            loading: false, 
            channel: {},
            channelName: 'test-channel'
        })
        expect(wrapper).toMatchSnapshot();
    })
})