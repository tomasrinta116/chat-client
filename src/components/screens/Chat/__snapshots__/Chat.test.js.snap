// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> it renders a loading spinner while loading 1`] = `
<div
  className="Chat-loadingSpinner-4"
>
  <WithStyles(CircularProgress) />
</div>
`;

exports[`<Chat /> it renders after loading 1`] = `
<Fragment>
  <WithStyles(Navbar)
    history={Object {}}
    sb={Object {}}
  />
  <div
    className="Chat-chatContainer-1"
  >
    <div
      className="Chat-chatHeader-3"
    >
      <ChatHeader
        title="test-channel"
      >
        <ChatMenu
          channel={Object {}}
          history={Object {}}
          participants={Array []}
          sb={Object {}}
        />
      </ChatHeader>
    </div>
    <div
      className="Chat-messagesDisplay-5"
    >
      <WithStyles(MessagesDisplay)
        messages={Array []}
      />
    </div>
    <div
      className="Chat-createMessage-2"
    >
      <WithStyles(MessageInput)
        addMessage={[Function]}
        channel={Object {}}
      />
    </div>
  </div>
</Fragment>
`;
